smolyay/smolyak.py

```diff
@@ -0,0 +1,157 @@
+
+        # cumulative sum to get end index of each level, up to exactness+1
+        end_levels = numpy.cumsum(self.index_per_level[:self.exactness+1])
+        # create ranges of indexes at each level
+        level_indexes = [list(range(end-n, end))
+                         for end, n in zip(end_levels, self.index_per_level)]
+        self._level_indexes = level_indexes
+
+        points = None
+        # get all combinations of points at each level
+        for summ in range(self.dimension, self.exactness+self.dimension+1):
+            # use generate_compositions function which implements NEXCOM
+            compositions = generate_compositions(summ, self.dimension)
+            for composition in compositions:
+                point_ids = [level_indexes[i] for i in composition]
+                points_ = numpy.array(
+                    numpy.meshgrid(*point_ids)).T.reshape(-1,
+                                                          self.dimension)
+                if points is None:
+                    points = points_
+                else:
+                    points = numpy.concatenate((points, points_), axis=0)
+        self._grid_point_index = points
+        self._needs_update = False
+
+    @property
+    def level_indexes(self):
+        """Generate indexes of levels.
+
+        Returns
+        -------
+        level_indexes : list
+            Unique indexes at each level.
+
+        """
+        if self._needs_update:
+            self._update()
+
+        return self._level_indexes
+
+    @property
+    def grid_point_index(self):
+        """Generate index of grid points.
+
+        It generates allowed compositons of levels,
+        and depending on the indexes for each level,
+        it makes the indexes of the grid points.
+
+        Returns
+        -------
+        grid_points_index : numpy array
+            Index of grid points
+
+        """
+        if self._needs_update:
+            self._update()
+
+        return self._grid_point_index
+
+
+def generate_compositions(summ, dimension):
+    """Genearte allowed compositions of levels using NEXCOM algorithm.
+
+    Parameters
+    ----------
+    summ: int
+        Summation of levels.
+    dimension: int
+        Number of independent variables.
+
+    Returns
+    -------
+    compositions: list
+        Allowed composition of levels.
+    """
+    max_level_index = summ - dimension
+    num_output = scipy.special.comb(max_level_index+dimension-1,
+                                    dimension-1, exact=True)
+    compositions = numpy.zeros((num_output, dimension),
+                               dtype=numpy.int32)
+
+    # (A) first entry
+    r = numpy.zeros(dimension, dtype=numpy.int32)
+    r[0] = max_level_index
+    t = max_level_index
+    h = 0
+    compositions[0] = r
+    index = 1
+
+    # (D): these termination conditions should be redundant
+    while (r[dimension-1] != max_level_index
+           and index < num_output + 1):
```

Comment:
I think the `+1` is incorrect here. I would also only use one check or the other, as these are redundant. Technically, it would be faster to use only the first one, since then you don't have to precompute the number of compositions.